File: api/routers/admin/categories.py
from fastapi import APIRouter, Depends, Request
from fastapi.encoders import jsonable_encoder
from fastapi.responses import JSONResponse
from database.engine import Database
from api.models.category import CategoryRead, CategoryCreate, CategoryUpdate, SearchCreate, SearchUpdate
from api.routers.access_token import check_admin_token

r = APIRouter(prefix="", tags=["Admin"])
db = Database()


@r.put("/update_category", response_model=CategoryRead)
async def update_category(category: CategoryUpdate):
    try:
        admin = await check_admin_token(request)
        if not admin:
            return JSONResponse(status_code=401, content={"message": "Unauthorized"})
        category = await db.update_category(id=category.id, name=category.name, image_path=category.image_path)
        return jsonable_encoder(CategoryRead.model_validate(category))
    except Exception as e:
        print(e)
        return JSONResponse(status_code=202, content=str(e))


@r.post("/add_search", response_model=CategoryRead)
async def add_search(search: SearchCreate, request: Request):
    try:
        admin = await check_admin_token(request)
        if not admin:
            return JSONResponse(status_code=401, content={"message": "Unauthorized"})
        search = await db.add_search(name=search.name, category_id=search.category_id,
                                     sub_category_id=search.sub_category_id)
        if search:
            category = await db.get_category(id=search.category_id)
            return jsonable_encoder(CategoryRead.model_validate(category))
    except Exception as e:
        print(e)
        return JSONResponse(status_code=202, content=str(e))

@r.put("/update_search", response_model=CategoryRead)
async def update_search(search: SearchUpdate, request: Request):
    try:
        category_id = search.category_id
        admin = await check_admin_token(request)
        if not admin:
            return JSONResponse(status_code=401, content={"message": "Unauthorized"})
        search = await db.update_search(id=search.id, name=search.name, category_id=search.category_id,
                                     sub_category_id=search.sub_category_id)
        if search:
            category = await db.get_category(id=category_id)
            return jsonable_encoder(CategoryRead.model_validate(category))
    except Exception as e:
        print(e)
        return JSONResponse(status_code=202, content=str(e))

@r.delete("/delete_search", response_model=CategoryRead)
async def delete_search(search_id: int, request: Request):
    try:
        admin = await check_admin_token(request)
        if not admin:
            return JSONResponse(status_code=401, content={"message": "Unauthorized"})
        category_id = await db.delete_search(id=search_id)
        if category_id:
            category = await db.get_category(id=category_id)
            return jsonable_encoder(CategoryRead.model_validate(category))
    except Exception as e:
        print(e)
        return JSONResponse(status_code=202, content=str(e))

File: admin_panel/src/requests/categories.ts
import axios from "axios";
import {CategoryRead, SearchCreate, SearchUpdate, SubCatgeoryCreate, SubCatgeoryUpdate} from "../models.ts";
import {url} from "./url";

export const getCategories = async (): Promise<CategoryRead[]> => {
  const {data, status} = await axios.get<CategoryRead[]>(url + '/categories')
  if (status !== 200) {
    return []
  }
  return data
}

export const updateCategory = async (category: CategoryRead): Promise<CategoryRead> => {
  const {data, status} = await axios.put<CategoryRead>(url + '/admin/update_category', category, {withCredentials: true})
  if (status !== 200) {
    return category
  }
  return data
}

export const addSearch = async (search: SearchCreate): Promise<CategoryRead | undefined> => {
  const {data, status} = await axios.post<CategoryRead>(url + '/admin/add_search', search, {withCredentials: true})
  if (status === 200) {
    return data
  }
}


export const updateSearch = async (search: SearchUpdate): Promise<CategoryRead | undefined> => {
  const {data, status} = await axios.put<CategoryRead>(url + '/admin/update_search', search, {withCredentials: true})
  if (status === 200) {
    return data
  }
}

export const deleteSearch = async (searchId: number): Promise<CategoryRead | undefined> => {
  const {data, status} = await axios.delete<CategoryRead>(url + '/admin/delete_search/?search_id='+searchId, {withCredentials: true})
  if (status === 200) {
    return data
  }
}

export const updateSubCategory = async (subCategory: SubCatgeoryUpdate, file: File | undefined): Promise<CategoryRead> => {
  const formData = new FormData()

  if (file) {
    formData.append('image_file', file)
  }
  formData.append('name', subCategory.name || '')
  formData.append('id', subCategory.id.toString())
  formData.append('is_available', subCategory.isAvailable?.toString() || '')

  const {data} = await axios.put<CategoryRead>(url + '/admin/update_sub_category', formData, {withCredentials: true, headers: {'Content-Type': 'multipart/form-data'}})
  return data
}

export const addSubCategory = async (subCategory: SubCatgeoryCreate, file: File | undefined): Promise<CategoryRead> => {
  const formData = new FormData()
  if (file) {
    formData.append('image_file', file)
  }
  formData.append('category_id', subCategory.categoryId.toString())
  formData.append('name', subCategory.name || '')
  formData.append('is_available', subCategory.isAvailable?.toString() || '')
  const {data} = await axios.post<CategoryRead>(url + '/admin/add_sub_category', formData, {withCredentials: true})
  return data
}

export const checkImage = async (name: string, host: string) => {
    const {status} = await axios.get(host + '/' + name, {withCredentials: true})
    if (status === 200) {
      return name
    }
    return ''
  }